### PR TITLE
feat(forms): introduce min and max validators

### PR DESCRIPTION
**Pull request by sonukapoor**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

File: packages/forms/src/directives/validators.ts [hunk 1]
```diff
@@ -193,6 +193,95 @@ export class RequiredValidator implements Validator {
...
+
+/**
+ * A directive which installs the {@link MinValidator} for any `formControlName`,
+ * `formControl`, or control with `ngModel` that also has a `min` attribute.
+ *
+ * @experimental
+ */
+@Directive({
+  selector:
+      '[formControlName],input[type=number][min][formControl],input[type=number][min][ngModel]',
```
Comment (AndrewKushnir) [inline packages/forms/src/directives/validators.ts:210]: I believe we should have the `input[type=number][min]` selector for the `[formControlName]` as well:
```suggestion
      'input[type=number][min][formControlName],input[type=number][min][formControl],input[type=number][min][ngModel]',
```

File: packages/forms/src/directives/validators.ts [hunk 2]
```diff
@@ -193,6 +193,95 @@ export class RequiredValidator implements Validator {
...
+
+/**
+ * A directive which installs the {@link MaxValidator} for any `formControlName`,
+ * `formControl`, or control with `ngModel` that also has a `min` attribute.
+ *
+ * @experimental
+ */
+@Directive({
+  selector:
+      '[formControlName],input[type=number][max][formControl],input[type=number][max][ngModel]',
```
Comment (AndrewKushnir) [inline packages/forms/src/directives/validators.ts:255]:
```suggestion
      'input[type=number][max][formControlName],input[type=number][max][formControl],input[type=number][max][ngModel]',
```

File: packages/forms/src/directives/validators.ts [hunk 3]
```diff
@@ -193,6 +193,95 @@ export class RequiredValidator implements Validator {
...
+  providers: [MIN_VALIDATOR],
+  host: {'[attr.min]': 'min ? min : null'}
+})
+export class MinValidator implements Validator, OnChanges {
+  private _validator: ValidatorFn;
+  private _onChange: () => void;
+
+  @Input() min: string;
+
+  ngOnChanges(changes: SimpleChanges): void {
```
Comment (AndrewKushnir) [inline packages/forms/src/directives/validators.ts:220]: We should add the `@nodoc` annotations for the `ngOnChanges`, `validate` and `registerOnValidatorChange` methods to exclude them from appearing on angular.io website (the same comment applies to the `MaxValidator`).

**Review by AndrewKushnir — COMMENTED** [review]
Thanks for creating this PR @sonukapoor, I've just left a couple comments.

**Comment by AndrewKushnir** [comment]
Hi @sonukapoor, I know this PR is in progress, but I'd like to share one idea that we can explore. The idea is to extract the common validator directive logic (that is repeated in all existing validators and the ones that you're adding as a part of this PR) to a base class, please see some code below. Note: the code is to illustrate the idea and it needs cleanup and a review of types, etc.

The base class may look like this:

```
/**
 * A base class for Validator-based Directives. The class contains common logic shared across such Directives.
 *
 * For internal use only, this class is not expected to be used outside of the Forms package.
 *
 * @experimental
 */
@Directive()
export abstract class AbstractValidatorDirective implements Validator, OnChanges {
  private _validator: ValidatorFn = Validators.nullValidator;
  private _onChange!: () => void;

  /**
   * Name of an input that matches directive selector attribute (e.g. `minlength` for `MinLengthDirective`).
   * An input with a given name might contain configuration information (like `minlength=“10”`) or a flag that
   * indicates whether validator should be enabled (like `[required]=“false”`).
   *
   * @nodoc
   */
  abstract inputName: string;

  /**
   * Creates an instance of a validator (specific to a directive that extends this base class). 
   *
   * @nodoc
   */
  abstract createValidator(input: unknown): ValidatorFn;

  /**
   * Performs the necessary input normalization based on a specific logic of a Directive.
   * For example, the function might be used to convert string-based representation of the `minlength`
   * input to an integer value that can later be used in the `Validators.minLength` validator.
   * 
   * @nodoc
   */
  abstract normalizeInput(input: unknown): unknown;

  /** @nodoc */
  ngOnChanges(changes: SimpleChanges): void {
    if (this.inputName in changes) {
      const input = this.normalizeInput(changes[this.inputName]);
      this._validator = this._createValidator(input);
      if (this._onChange) {
        this._onChange();
      }
    }
  }

  /** @nodoc */
  validate(control: AbstractControl): ValidationErrors|null {
    return this._validator(control);
  }

  /** @nodoc */
  registerOnValidatorChange(fn: () => void): void {
    this._onChange = fn;
  }
}
```

The `MaxValidator` directive extends the base one: 
```
/**
 * A directive which installs the {@link MaxValidator} for any `formControlName`,
 * `formControl`, or control with `ngModel` that also has a `max` attribute.
 *
 * @experimental
 */
@Directive({
  selector: ‘input[type=number][max][formControlName],’ +
                  ‘input[type=number][max][formControl],’ +
                  ‘input[type=number][max][ngModel]',
  providers: [MAX_VALIDATOR],
  host: {'[attr.max]': 'max ? max : null'}
})
export class MaxValidator extends AbstractValidatorDirective {
  @Input() max!: string;

  /* @internal */
  inputName = ’max’;

  /* @internal */
  normalizeInput = (input: string): number => parseInt(input, 10);

  /* @internal */
  createValidator = (max: number): ValidatorFn => Validators.max(max);
}
```

Also this is how existing validators may look like (for ex. the `RequiredValidator` one):
```
// Utility function that converts directive input to a boolean value.
const toBoolean = (input: string|null): boolean => input != null && input !== false && `${input}` !== 'false’;

@Directive({
  selector: ':not([type=checkbox])[required][formControlName],’ +
                  ‘:not([type=checkbox])[required][formControl],’ +
                  ‘:not([type=checkbox])[required][ngModel]',
  providers: [REQUIRED_VALIDATOR],
  host: {'[attr.required]': 'required ? "" : null'}
})
export class RequiredValidator extends AbstractValidatorDirective {
  @Input() required!: string|null;

  /* @internal */
  inputName = ’required’;

  /* @internal */
  normalizeInput = toBoolean;

  /* @internal */
  createValidator = (enabled: boolean): ValidatorFn => enabled ? Validators.required : Validators.nullValidator;
}
```

I think we should consider adding the base class and use it for 2 new directives. Existing ones can be switched over later once we confirm the overall approach.

What do you think?

**Comment by sonukapoor** [comment]
> What do you think?

@AndrewKushnir I think that makes sense. As discussed, I will close this PR and start fresh work on this. 👍

**Pull request closed without merging**